R/filters.R

```diff
@@ -74,3 +116,200 @@ filter_by_meta <- function(.data, .query) {
+          counts < as_numeric_or_fail(query_args)
+        ]
+      } else if (query_type == "morethan") {
+        filtered_data <- filtered_data[
+          counts > as_numeric_or_fail(query_args)
+        ]
+      } else if (query_type == "interval") {
+        from <- as_numeric_or_fail(query_args[[1]])
+        to <- as_numeric_or_fail(query_args[[2]])
+        filtered_data <- filtered_data[
+          counts >= from & counts < to
+        ]
+      }
+
+      if (length(filtered_data) == 0) {
+        warning(paste0("Filter by key \"", name, "\" removed all remaining samples!"))
+      }
+    }
+  }
+  filtered_meta <- subset(.data$meta, Sample %in% names(filtered_data))
+
+  return(list(data = filtered_data, meta = filtered_meta))
+}
+
+filter_by_clonotype <- function(.data, .query, .match) {
+  filtered_data <- .data$data
+  for (i in seq_along(names(.query))) {
+    name <- names(.query)[i]
+    column_query <- .query[[name]]
+    query_type <- column_query[[1]]
+    query_args <- column_query[-1]
+    for (j in seq_along(names(.data$data))) {
+      sample_name <- names(.data$data)[j]
+      sample <- filtered_data[[sample_name]]
+      if (!(name %in% names(sample))) {
+        stop(paste0("Column \"", name, "\" not found in sample \"", sample_name, "\"."))
+      }
+      if (nrow(sample) == 0) {
+        warning(paste0("Sample \"", sample_name, "\" was empty, removed!"))
+      } else {
+        sample %<>% filter_table(name, query_type, query_args, .match)
+        if (nrow(sample) == 0) {
+          warning(paste0(
+            "Filter by column \"", name,
+            "\" removed all remaining clonotypes from sample \"",
+            sample_name, "\"; sample was removed!"
+          ))
+        }
+      }
+
+      if (nrow(sample) == 0) {
+        # removing empty sample
+        filtered_data <- filtered_data[names(filtered_data) != sample_name]
+      } else {
+        # updating sample in filtered_data
+        filtered_data[[sample_name]] <- sample
+      }
+    }
+  }
+  filtered_meta <- subset(.data$meta, Sample %in% names(filtered_data))
+
+  return(list(data = filtered_data, meta = filtered_meta))
+}
+
+filter_table <- function(.table, .column_name, .query_type, .query_args, .match) {
+  if (.query_type == "include") {
+    if (.match == "exact") {
+      .table %<>% subset(get(.column_name) %in% .query_args)
+    } else if (.match == "startswith") {
+      .table <- .table[
+        unlist(lapply(lapply(.table[[.column_name]], startsWith, .query_args), any)),
+      ]
+    } else if (.match == "substring") {
+      .table <- .table[
+        unique(unlist(lapply(.query_args, grep, .table[[.column_name]], fixed = TRUE))),
+      ]
+    }
+  } else if (.query_type == "exclude") {
+    if (.match == "exact") {
+      .table %<>% subset(!get(.column_name) %in% .query_args)
+    } else if (.match == "startswith") {
+      .table <- .table[
+        !unlist(lapply(lapply(.table[[.column_name]], startsWith, .query_args), any)),
```

Comment:
starts_with I think is better to use instead of multiple apply, maybe? something like `starts_with(match = .query_args, vars = .table[[.column_name]], ignore.case=F)` will return you index of matched with query objects;